hail/src/main/scala/is/hail/expr/ir/InferPType.scala

```diff
@@ -8,65 +8,140 @@ object InferPType {
+      case Str(_) => PString(true)
+      case Literal(t, v) => PType.canonical(t, v != null)
+      case True() | False() => PBoolean(true)
       case Void() => PVoid
-      case Cast(_, t) => PType.canonical(t)
-      case CastRename(_, t) => PType.canonical(t)
-      case NA(t) => PType.canonical(t)
-      case IsNA(_) => PBoolean()
+      case Cast(ir, t) => {
+        InferPType(ir, env)
+        PType.canonical(t, ir.pType2.required)
+      }
+      case CastRename(ir, t) => {
+        InferPType(ir, env)
+        PType.canonical(t, ir.pType2.required)
+      }
+      case NA(t) => PType.canonical(t, false)
+      case IsNA(ir) => {
+        InferPType(ir, env)
+        PBoolean(true)
+      }
       case Coalesce(values) => {
         val vit = values.iterator
         val head = vit.next()
         InferPType(head, env)
 
+        var allRequired = head.pType2.required
         while(vit.hasNext) {
           val value = vit.next()
           InferPType(value, env)
-          assert(head.pType2 == value.pType2)
-        }
+          assert(head.pType2 isOfType value.pType2)
 
-        head.pType2
+          if(allRequired == true && value.pType2.required == false) {
+            allRequired = false
+          }
+         }
+
+        head.pType2.setRequired(allRequired)
       }
       case Ref(name, _) => env.lookup(name)
       case In(_, t) => PType.canonical(t)
-      case MakeArray(_, t) => PType.canonical(t)
-      case MakeStream(_, t) => PType.canonical(t)
+      case MakeArray(irs, _) => {
+        val it = irs.iterator
+        val head = it.next()
+        InferPType(head, env)
+
+        var allRequired = head.pType2.required
+        while(it.hasNext) {
+          val irElem = it.next()
+
+          InferPType(irElem, env)
+          assert(head.pType2 isOfType irElem.pType2)
+
+          if(allRequired == true && irElem.pType2.required == false) {
+            allRequired = false
+          }
+        }
+
+        PArray(head.pType2.setRequired(allRequired), true)
+      }
+      case MakeStream(irs, _) => {
+        val it = irs.iterator
+        val head = it.next()
+        InferPType(head, env)
+
+        var allRequired = head.pType2.required
+        while(it.hasNext) {
+          val irElem = it.next()
+
+          InferPType(irElem, env)
+          assert(head.pType2 isOfType irElem.pType2)
+
+          if(allRequired == true && irElem.pType2.required == false) {
+            allRequired = false
+          }
+        }
+
+        PStream(head.pType2.setRequired(allRequired), true)
+      }
       case MakeNDArray(data, shape, _) => {
         InferPType(data, env)
         InferPType(shape, env)
 
         val nElem = shape.pType2.asInstanceOf[PTuple].size
+        val etype = coerce[PArray](data.pType2).elementType
 
-        PNDArray(coerce[PArray](data.pType2).elementType.setRequired(true), nElem)
+        assert(etype.required == true)
+        PNDArray(etype, nElem, data.pType2.required)
```

Comment:
One thing, in merge, there was a conflict at this line, was
`PNDArray(coerce[PArray](data.pType2).elementType.setRequired(true), nElem)`

There was a merge conflict, where this line read
`PNDArray(coerce[PArray](data.pType2).elementType.setRequired(true), nElem)`

I removed the setRequired true, because I assert that the `elementType` is required. I think this is better, because it relies on the caller to do the right thing, rather than providing a different type than they request. I can imagine that they could pass nulls while we assert true, and we get a mismatch between the actual emitted values and the downstream expectation.

Reply:
the NDArray nodes check that elements are not null, and error if any are. It's incorrect to remove the setRequired because the data IR could be optional, but the element type of the NDArray must be required.

Reply:
Does this not cover that case?

```scala
val etype = coerce[PArray](data.pType2).elementType
assert(etype.required == true)
```

Reply:
the null check needs to be at runtime, not at compile time. It's perfectly valid for a user to pass data with an optional element type into MakeNDArray, because it's possible that the user knows that there are no elements that are missing, even if Hail can't prove that statically.

Reply:
Addressed, along with fix for NDArrayShape, in spirit with ArrayRange.